pennylane/qnodes/passthru.py

```diff
@@ -0,0 +1,145 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+PassthruQNode class
+"""
+import pennylane as qml
+import pennylane.operation
+import pennylane.circuit_graph
+from .base import BaseQNode, QuantumFunctionError
+
+
+
+class PassthruQNode(BaseQNode):
+    """Differentiable quantum node that appears as a white box to an external autodiff framework.
+
+    In PennyLane, the QNode classes work as black box functions with respect to any
+    autodiff (AD) framework (such as TensorFlow or PyTorch). This means that the QNode
+    converts all its inputs (which may come in data types specific to the
+    AD framework used, which we denote *ADT* here) into plain Python/NumPy types,
+    computes the required :ref:`quantum function <intro_vcirc_qfunc>` value or Jacobian,
+    and converts the result back into the ADT if necessary.
+
+    In contrast, PassthruQNode works as a white box: it preserves the ADT
+    throughout the computation. This requires that the quantum function is computed
+    using a simulator device that is compatible with the AD framework used (typically
+    implemented using that same framework).
+
+    The advantages of this approach is that the qfunc can be differentiated using its AD framework
+    without requiring a separate method for computing the Jacobian, and that the internals
+    of the simulation are visible in the computational graph.
+
+    PassthruQNode requires some changes to the way other PennyLane components work:
+
+    1. :class:`Operator` must not do domain checking for its parameters, or it must let the ADT pass the check.
+    2. The simulator device must return the result as the ADT instead of plain Python/NumPy types.
+    3. Any output_conversion in :meth:`BaseQNode.evaluate` must be skipped.
```

Comment:
Other than #2 (which is a requirement for supported simulators), these other points don't belong in the docstring. Instead, the changes should just be made in the code and documented in the PR

Reply:
These are just design notes, I wanted the reviewers to read them before I clean up the docstring.

Reply:
Still good habit to leave as comments rather than docstrings, since it's possible we'll forget to remove them in future PRs :laughing: